src/v/serde/serde.h

```diff
@@ -243,213 +238,298 @@ int64_t checked_duration_cast_to_nanoseconds(
+}
+
+template<typename T>
+requires(
+  std::is_scalar_v<std::decay_t<
+    T>> && !serde_is_enum_v<std::decay_t<T>>) void write(iobuf& out, T t) {
     using Type = std::decay_t<T>;
+    if constexpr (sizeof(Type) == 1) {
+        out.append(reinterpret_cast<char const*>(&t), sizeof(t));
+    } else if constexpr (std::is_same_v<float, Type>) {
+        auto const le_t = htole32(bit_cast<uint32_t>(t));
+        static_assert(sizeof(le_t) == sizeof(Type));
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(le_t));
+    } else if constexpr (std::is_same_v<double, Type>) {
+        auto const le_t = htole64(bit_cast<uint64_t>(t));
+        static_assert(sizeof(le_t) == sizeof(Type));
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(le_t));
+    } else {
+        auto const le_t = ss::cpu_to_le(t);
+        static_assert(sizeof(le_t) == sizeof(Type));
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(le_t));
+    }
+}
+
+template<typename T>
+requires(serde_is_enum_v<std::decay_t<T>>) void write(iobuf& out, T t) {
+    using Type = std::decay_t<T>;
+    auto const val = static_cast<std::underlying_type_t<Type>>(t);
+    if (unlikely(
+          val > std::numeric_limits<serde_enum_serialized_t>::max()
+          || val < std::numeric_limits<serde_enum_serialized_t>::min())) {
+        throw serde_exception{fmt_with_ctx(
+          ssx::sformat,
+          "serde: enum of type {} has value {} which is out of bounds for "
+          "serde_enum_serialized_t",
+          type_str<T>(),
+          val)};
+    }
+    write(out, static_cast<serde_enum_serialized_t>(val));
+}
+
+template<typename Rep, typename Period>
+void write(iobuf& out, std::chrono::duration<Rep, Period> t) {
+    // We explicitly serialize it as ns to avoid any surprises like
+    // seastar updating underlying duration types without
+    // notice. See https://github.com/redpanda-data/redpanda/pull/5002
+    //
+    // Check for overflows/underflows.
+    // For ex: a millisecond and nanosecond use the same underlying
+    // type int64_t but converting from one to other can easily overflow,
+    // this is by design.
+    // Since we serialize with ns precision, there is a restriction of
+    // nanoseconds::max()'s equivalent on the duration to be serialized.
+    // On a typical platform which uses int64_t for 'rep', it roughly
+    // translates to ~292 years.
+    //
+    // If we detect an overflow, we will clamp it to maximum supported
+    // duration, which is nanosecond::max() and if there is an underflow,
+    // we clamp it to minimum supported duration which is nanosecond::min().
     static_assert(
-      !is_chrono_time_point<Type>,
-      "Time point serialization is risky and can have unintended "
-      "consequences. Check with Redpanda team before fixing this.");
-    static_assert(are_bytes_and_string_different<Type>);
-    static_assert(has_serde_write<Type> || is_serde_compatible_v<Type>);
```

Comment:
Is there a reason these two static asserts were left out? I can recall the second one being especially useful

Reply:
`are_bytes_and_string_different` should not be needed anymore since now we have two overloads for `serde::write`:

  - one for `bytes`
  - one for `string`

So if you would have the case that `bytes` and `string` are exactly the same type, the `write` function would be a duplicate - so the compiler will catch this case and won't compile the code.

Reply:
The check `has_serde_write<Type> || is_serde_compatible_v<Type>` is not needed anymore because we don't have one big function anymore with a chain of `else if constexpr (... /* type detection */)`.

Instead we have one overload for each type. So the case `!has_serde_write<Type>` will automatically be caught by the compiler when it does not find a fitting overload.